src/Avalonia.Base/Media/TextFormatting/TextLineImpl.cs

```diff
@@ -1054,191 +1076,157 @@ private TextBounds GetTextBoundsLeftToRight(int firstRunIndex, int lastRunIndex,
+                var previousCharacterHit =
+                    currentRun.GlyphRun.GetPreviousCaretCharacterHit(new CharacterHit(startIndex - runOffset));
+                var nextCharacterHit = currentRun.GlyphRun.GetNextCaretCharacterHit(previousCharacterHit);
 
-                //We need to move the startIndex to the start of the cluster.
-                startIndex -= clusterOffset;
-            }
-
-            var characterLength = Math.Max(0, Math.Abs(startHit.FirstCharacterIndex + startHit.TrailingLength - 
-                endHit.FirstCharacterIndex - endHit.TrailingLength) - clusterOffset);
-
-            var runOffset = startIndex - offset;
-
-            if (characterLength == 0 && currentRun.Text.Length > 0 && runOffset < currentRun.Text.Length)
-            {
-                //Make sure we are properly dealing with zero width space runs
-                var codepointEnumerator = new CodepointEnumerator(currentRun.Text.Span.Slice(runOffset));
+                var clusterLength =
+                    nextCharacterHit.FirstCharacterIndex + nextCharacterHit.TrailingLength -
+                    (previousCharacterHit.FirstCharacterIndex + previousCharacterHit.TrailingLength);
 
-                while (remainingLength > 0 && codepointEnumerator.MoveNext(out var codepoint))
+                if (clusterLength > runOffset)
                 {
-                    if (codepoint.IsWhiteSpace)
-                    {
-                        characterLength++;
-                        remainingLength--;
-                    }
-                    else
-                    {
-                        break;
-                    }
+                    //Remember the cluster correction offset
+                    clusterOffset = clusterLength - runOffset;
+
+                    //Move to the start of the cluster
+                    startIndex -= clusterOffset;
                 }
             }
 
-            if (startHit.FirstCharacterIndex > endHit.FirstCharacterIndex)
-            {
-                startHit = endHit;
-            }
+            //Find the visual start and end position we want to hit test against
+            var startOffset = currentRun.GlyphRun.GetDistanceFromCharacterHit(new CharacterHit(startIndex));
+            var endOffset = currentRun.GlyphRun.GetDistanceFromCharacterHit(new CharacterHit(endIndex));
 
-            if (endX < startX)
+            // Preserve non-zero width for zero-advance ranges
+            if (startOffset == endOffset && startIndex != endIndex)
             {
-                (endX, startX) = (startX, endX);
+                //We need to make sure a zero width text line is hit test at the end so we add some delta
+                endOffset += MathUtilities.DoubleEpsilon;
             }
+        
+            // Map offsets to visual coordinates using the provided endX (right edge)
+            var visualStartX = endX - (currentRun.Size.Width - startOffset);
```

Comment:
Question: why do we have a notion of visual bounds (`visualStartX`/`visualEndX`) in RTL mode but not in LTR? Is that a possible logic mistake or simply something normal?

Can you please make the variables names match as much as possible between the two modes?